src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.cs

```diff
@@ -162,72 +123,98 @@ private void SyntaxNodeAction(SyntaxNodeAnalysisContext syntaxContext)
-                localDeclarationStatement.GetLocation(),
-                ifStatement.GetLocation(),
-                condition.GetLocation(),
-                initializerValue.GetLocation());
+                declarationStatement.GetLocation(),
+                ifOrWhileStatement.GetLocation(),
+                leftmostCondition.GetLocation(),
+                asExpression.GetLocation());
 
             // Put a diagnostic with the appropriate severity on the declaration-statement itself.
             syntaxContext.ReportDiagnostic(Diagnostic.Create(
-                GetDescriptorWithSeverity(severity),
-                localDeclarationStatement.GetLocation(),
+                GetDescriptorWithSeverity(styleOption.Notification.Value),
+                declarationStatement.GetLocation(),
                 additionalLocations));
         }
 
-        private bool IsAccessedBeforeAssignment(
-            SyntaxNodeAnalysisContext syntaxContext,
-            VariableDeclaratorSyntax declarator,
-            IfStatementSyntax ifStatement,
+        private static bool CheckDefiniteAssignment(
+            SemanticModel semanticModel,
+            ISymbol localVariable,
+            StatementSyntax declarationStatement,
+            StatementSyntax usageStatement,
+            SyntaxList<StatementSyntax> blockStatements,
             CancellationToken cancellationToken)
         {
-            var semanticModel = syntaxContext.SemanticModel;
-            var localVariable = semanticModel.GetDeclaredSymbol(declarator);
-
             var isAssigned = false;
             var isAccessedBeforeAssignment = false;
 
-            CheckDefiniteAssignment(
-                semanticModel, localVariable, ifStatement.Else,
-                out isAssigned, out isAccessedBeforeAssignment,
-                cancellationToken);
+            var usageIndex = blockStatements.IndexOf(usageStatement);
+            var declarationIndex = blockStatements.IndexOf(declarationStatement);
 
-            if (isAccessedBeforeAssignment)
+            // Since we're going to remove this declaration-statement, 
+            // we need to first ensure that it's not used up to the target statement.
+            for (var index = declarationIndex + 1; index < usageIndex; index++)
             {
-                return true;
+                CheckDefiniteAssignment(
+                    semanticModel, localVariable, blockStatements[index],
+                    out isAssigned, out isAccessedBeforeAssignment,
+                    cancellationToken);
+
+                if (isAssigned || isAccessedBeforeAssignment)
+                {
+                    return false;
+                }
             }
 
-            var parentBlock = (BlockSyntax)ifStatement.Parent;
-            var ifIndex = parentBlock.Statements.IndexOf(ifStatement);
-            for (int i = ifIndex + 1, n = parentBlock.Statements.Count; i < n; i++)
+            var @while = true;
```

Comment:
just remove this variable, or just initialize it right here.  It is really weird to say it's a while, and then undo that once you see that it's an if.